frontend/src/metabase/actions/containers/ActionCreatorModal/ActionCreatorModal.unit.spec.tsx

```diff
@@ -102,10 +115,212 @@ describe("actions > containers > ActionCreatorModal", () => {
+
+      await waitFor(() => {
+        expect(screen.getByTestId("action-creator")).toBeInTheDocument();
+      });
+
+      userEvent.type(screen.getByDisplayValue("New Action"), "a change");
+      userEvent.tab(); // need to click away from the input to trigger the isDirty flag
+
+      history.goBack();
+
+      expect(screen.getByText("Changes were not saved")).toBeInTheDocument();
+      expect(
+        screen.getByText(
+          "Navigating away from here will cause you to lose any changes you have made.",
+        ),
+      ).toBeInTheDocument();
+    });
+
+    it("does not show custom warning modal when saving changes", async () => {
+      const initialRoute = `/model/${MODEL.id}/detail/actions`;
+      const actionRoute = `/model/${MODEL.id}/detail/actions/new`;
+      const { history } = await setup({ initialRoute, action: null });
+
+      history.push(actionRoute);
+
+      await waitFor(() => {
+        expect(screen.getByTestId("action-creator")).toBeInTheDocument();
+      });
+
+      const query = "select 1;";
+
+      userEvent.type(screen.getByDisplayValue("New Action"), "a change");
+      userEvent.type(screen.queryAllByRole("textbox")[1], query);
+      userEvent.tab(); // need to click away from the input to trigger the isDirty flag
+
+      fetchMock.post("path:/api/action", {
+        name: "New Actiona change",
+        dataset_query: {
+          type: "native",
+          database: DATABASE.id,
+          native: {
+            query,
+            "template-tags": {},
+          },
+        },
+        database_id: DATABASE.id,
+        parameters: [],
+        type: "query",
+        visualization_settings: {
+          name: "",
+          type: "button",
+          description: "",
+          confirmMessage: "",
+          successMessage: "",
+          fields: {},
+        },
+      });
+
+      userEvent.click(screen.getByRole("button", { name: "Save" }));
+      userEvent.click(screen.getByRole("button", { name: "Create" }));
+
+      await waitFor(() => {
+        expect(history.getCurrentLocation().pathname).toBe(initialRoute);
+      });
+
+      expect(
+        screen.queryByText("Changes were not saved"),
+      ).not.toBeInTheDocument();
+      expect(
+        screen.queryByText(
+          "Navigating away from here will cause you to lose any changes you have made.",
+        ),
+      ).not.toBeInTheDocument();
+    });
+  });
+
+  describe("editing existing action", () => {
+    it("does not show custom warning modal when leaving with no changes via SPA navigation", async () => {
+      const action = ACTION;
+      const initialRoute = `/model/${MODEL.id}/detail/actions`;
+      const actionRoute = `/model/${MODEL.id}/detail/actions/${action.id}`;
+      const { history } = await setup({ initialRoute, action });
+
+      history.push(actionRoute);
+
+      await waitFor(() => {
+        expect(screen.getByTestId("action-creator")).toBeInTheDocument();
+      });
+
+      history.goBack();
```

Comment:
bonus point: it'd be nice to test that if after you modify an existing action, you modify it back to the initial state e.g. 

- change the name
- <kbd>Tab</kbd>
- change the name back
- <kbd>Tab</kbd>

And the warning modal shouldn't be shown. This works because you're comparing both question and form settings' values.

Reply:
Good call :+1: f4b140caee357afe5432c9b30f9a64fe0c050de0